core/config/options.go

```diff
@@ -71,6 +91,56 @@ func (opts *Options) List(ctx context.Context, key string) ([][]string, error) {
 	return tuples, nil
 }
 
+// ListFunc returns a closure that returns the set of tuples for the
+// provided key.
+//
+// The configuration option for key must be a set of tuples.
+// ListFunc will panic if the provided key is undefined in the schema or
+// is defined as a scalar.
+//
+// The returned function will perform a stale read of the
+// configuration value. If an error occurs while reading the value,
+// the old value is returned and the error is saved on the Options
+// type to be returned in Err.
+func (opts *Options) ListFunc(key string) func() [][]string {
+	opt, ok := opts.schema[key]
+	if !ok {
+		panic(fmt.Errorf("unknown config option %q", key))
+	} else if opt.equalFunc == nil {
+		panic(fmt.Errorf("config option %q is a scalar, not a set", key))
+	}
+
+	// old is the last successfully retrieved tuple set for this
+	// configuration key. The returned closure will update it on
+	// every successful lookup. If an error occurs, the closure
+	// returns this value.
+	var old atomic.Value
+	old.Store([][]string(nil))
```

Comment:
This ensure that `old.Load().([][]string)` won't panic if the first call to the closure fails to load the value. Otherwise `old.Load()` would return an untyped nil which cannot be asserted to [][]string.

Alternatively, we could return an error from ListFunc and perform a synchronous lookup before returning the closure so that we always have a previous value available in the closure. It's hard to tell how onerous returning an error from ListFunc would be since we don't have any call sites yet.